### PR TITLE
fix: cluster-wide resources do not use `ownerRef`s

### LINKED ISSUE #72: `ControlPlane`'s `ClusterRole`, `ClusterRoleBinding` and `ValidatingWebhookConfiguration` incorrectly use namespaced resources as their owners
## Current Behavior

Cluster scoped resources like `ControlPlane`'s `ClusterRole` and `ClusterRoleBinding` incorrectly use namespaced resources - `ControlPlane`s - as their owner which create `OwnerRefInvalidNamespace` events:

```
apiVersion: v1
count: 1
eventTime: null
firstTimestamp: "2024-03-29T12:36:40Z"
involvedObject:
  apiVersion: admissionregistration.k8s.io/v1
  kind: ValidatingWebhookConfiguration
  name: kong-jw255
  uid: 8ee835da-51c1-4e7c-89ab-8ea24087f7b7
kind: Event
lastTimestamp: "2024-03-29T12:36:40Z"
message: 'ownerRef [gateway-operator.konghq.com/v1beta1/ControlPlane, namespace: ,
  name: kong-jw255, uid: 19de2c4c-f926-437b-8423-96efab88b63a] does not exist in namespace
  ""'
metadata:
  creationTimestamp: "2024-03-29T12:36:40Z"
  name: kong-jw255.17c13c739e2797e9
  namespace: default
  resourceVersion: "1295539"
  uid: d91d9ad7-22f4-40e1-9e31-2cd836063835
reason: OwnerRefInvalidNamespace
reportingComponent: garbage-collector-controller
reportingInstance: ""
source:
  component: garbage-collector-controller
type: Warning
```

```
apiVersion: v1
count: 1
eventTime: null
firstTimestamp: "2024-03-29T12:36:40Z"
involvedObject:
  apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRole
  name: kong-jw255-rcqfq
  uid: 9c35bcef-fbcd-4dcd-a2f0-7edb69414a6f
kind: Event
lastTimestamp: "2024-03-29T12:36:40Z"
message: 'ownerRef [gateway-operator.konghq.com/v1beta1/ControlPlane, namespace: ,
  name: kong-jw255, uid: 19de2c4c-f926-437b-8423-96efab88b63a] does not exist in namespace
  ""'
metadata:
  creationTimestamp: "2024-03-29T12:36:40Z"
  name: kong-jw255-rcqfq.17c13c739ce28698
  namespace: default
  resourceVersion: "1295524"
  uid: 3feaad08-4c6f-413f-a3fb-4c3b42c98e90
reason: OwnerRefInvalidNamespace
reportingComponent: garbage-collector-controller
reportingInstance: ""
source:
  component: garbage-collector-controller
type: Warning
```

```
apiVersion: v1
count: 1
eventTime: null
firstTimestamp: "2024-03-29T12:36:40Z"
involvedObject:
  apiVersion: rbac.authorization.k8s.io/v1
  kind: ClusterRoleBinding
  name: controlplane-kong-jw255-zbx6n
  uid: 79e0727e-56d8-465f-ad5d-605d3002b67b
kind: Event
lastTimestamp: "2024-03-29T12:36:40Z"
message: 'ownerRef [gateway-operator.konghq.com/v1beta1/ControlPlane, namespace: ,
  name: kong-jw255, uid: 19de2c4c-f926-437b-8423-96efab88b63a] does not exist in namespace
  ""'
metadata:
  creationTimestamp: "2024-03-29T12:36:40Z"
  name: controlplane-kong-jw255-zbx6n.17c13c739d0da343
  namespace: default
  resourceVersion: "1295526"
  uid: 0f73bd62-cd16-4b66-b115-865cbf8c1b0e
reason: OwnerRefInvalidNamespace
reportingComponent: garbage-collector-controller
reportingInstance: ""
source:
  component: garbage-collector-controller
type: Warning
```

## Expected Behavior

Cluster scoped resources shouldn't set namespaced resources as their owners.

## Steps To Reproduce

1. Deploy the operator
2. Deploy one of the manifests that will result in `ControlPlane` being deployed e.g. https://github.com/Kong/gateway-operator/blob/4e516dbbbc02d86f9601c7aed6182fc865b19cde/config/samples/gateway-with-gatewayconfiguration.yaml
3. Observe events being created

## Operator Version

4e516dbb

## Related materials

https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/

### PR DESCRIPTION
**Pull request by mlavacca**
**What this PR does / why we need it**:

**Which issue this PR fixes**

Fixes #72 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect significant changes

### CODE REVIEW

File: pkg/utils/kubernetes/ownerrefs.go [hunk 1]
```diff
@@ -38,6 +38,13 @@ func SetOwnerForObject(obj, owner client.Object) {
 	}
 }
 
+// SetOwnerForObjectThroughLabel sets the owner of the provided object through a label.
+func SetOwnerForObjectThroughLabel(obj, owner client.Object, label string) {
+	labels := obj.GetLabels()
+	labels[label] = owner.GetName()
```
Comment (czeslavo) [inline pkg/utils/kubernetes/ownerrefs.go]: Given that an owner might be a namespaced object (e.g. a ControlPlane) - aren't we losing here this information, putting only a name into the label? I think we also lose type information - we're not able to tell if an object labeled with `GatewayOperatorManagedByNameLabel` is a reference to a ControlPlane or some other type.

Comment (pmalek) [reply pkg/utils/kubernetes/ownerrefs.go]: 👍 

We could save the type (by checking the kind of the owner) and ns as well. Moreover the `label` is always `consts.GatewayOperatorManagedByNameLabel` so I'd drop this as an argument here.

When more types need to be supported we can always use generics and type asserts (or stick to `GetObjectKind()` and set the labels accordingly.

Comment (mlavacca) [reply pkg/utils/kubernetes/ownerrefs.go]: Yep, I've changed the code to use the following label set:

- `gateway-operator.konghq.com/managed-by`: `ControlPlane.gateway-operator.konghq.com`
- `gateway-operator.konghq.com/managed-by-name`: `<controlplane-name>`
- `gateway-operator.konghq.com/managed-by-namespace`: `<controlplane-namespace>`

Let me know what you think :)

Comment (pmalek) [reply pkg/utils/kubernetes/ownerrefs.go]: Regarding the `managed-by` label: currently we use this with just the uncapitalized kind name in the value, e.g. 

https://github.com/Kong/gateway-operator/blob/9015ff6fb46a2325bd5c63058d5c30bf58707e43/controller/controlplane/controller_reconciler_utils.go#L258

where

```
const (
	// ControlPlaneManagedLabelValue indicates that an object's lifecycle is managed
	// by the controlplane controller.
	ControlPlaneManagedLabelValue = "controlplane"
)
```

so the resource gets the following label:

```
gateway-operator.konghq.com/managed-by: gateway
```

I'd suggest that we stay consistent and keep it that way. WDYT?

Comment (mlavacca) [reply pkg/utils/kubernetes/ownerrefs.go]: sure thing, I've made it consistent 👍

File: pkg/utils/kubernetes/ownerrefs.go [hunk 2]
```diff
@@ -38,6 +42,40 @@ func SetOwnerForObject(obj, owner client.Object) {
...
+		return err
+	}
+	for k, v := range managedByLabelSet {
+		labels[k] = v
+	}
+	obj.SetLabels(labels)
+	return nil
+}
+
+// GetManagedByLabelSet returns a map of labels with the managing object's metadata.
```
Comment (pmalek) [inline pkg/utils/kubernetes/ownerrefs.go]:
```suggestion
// GetManagedByLabelSet returns a map of labels with the provided object's metadata.
// These can be applied to other objects that are owned by the object provided as an argument.
```

WDYT about this wording?

File: pkg/utils/kubernetes/ownerrefs.go [hunk 3]
```diff
@@ -38,6 +42,40 @@ func SetOwnerForObject(obj, owner client.Object) {
...
+func GetManagedByLabelSet(obj client.Object) (map[string]string, error) {
```
Comment (pmalek) [inline pkg/utils/kubernetes/ownerrefs.go]: - Food for thought: we could limit the types accepted by this func via generic type constraint and this way:
  
  - not loose type information
  - not need the `error` return type

- How about adding a simple UT for this to guard the functionality from changing unexpectedly?

File: pkg/utils/kubernetes/lists.go [hunk 4]
```diff
@@ -244,12 +224,11 @@ func ListSecretsForOwner(ctx context.Context,
 	return secrets, nil
 }
 
-// ListValidatingWebhookConfigurationsForOwner is a helper function to map a list of ValidatingWebhookConfiguration
-// by list options and reduce by OwnerReference UID to efficiently list only the objects owned by the provided UID.
-func ListValidatingWebhookConfigurationsForOwner(
+// ListValidatingWebhookConfigurations is a helper function to map a list of ValidatingWebhookConfiguration
+// by list options.
```
Comment (pmalek) [inline pkg/utils/kubernetes/lists.go]:
```suggestion
// ListValidatingWebhookConfigurations is a helper function which gets a list of ValidatingWebhookConfiguration
// using the provided list options.
```

Comment (pmalek) [reply pkg/utils/kubernetes/lists.go]: Could also be applied to other godocs.

File: pkg/utils/kubernetes/ownerrefs.go
Comment (pmalek) [inline pkg/utils/kubernetes/ownerrefs.go:1]: WDYT about adding a test case assertion to the e2e upgrade test which checks for the presence of those labels on the `ControlPlane`?

**Review by pmalek — COMMENTED** [review]
Couple of possible places where we can make things simpler + a proposal to add some tests.

👍

File: pkg/utils/kubernetes/ownerrefs.go (hunk 3 above)
Comment (mlavacca) [reply pkg/utils/kubernetes/ownerrefs.go]: yeah, it makes total sense. I've added the generics and a simple UT 👍

File: pkg/utils/kubernetes/ownerrefs.go
Comment (mlavacca) [reply pkg/utils/kubernetes/ownerrefs.go:1]: I've added it, let me know if it looks good to you :)

File: test/e2e/test_helm_install_upgrade.go [hunk 5]
```diff
@@ -466,3 +459,42 @@ func gatewayDataPlaneDeploymentIsNotPatched(gatewayLabelSelector string) func(co
...
+
+func clusterWideResourcesAreProperlyManaged(gatewayLabelSelector string) func(ctx context.Context, c *assert.CollectT, cl client.Client) {
+	return func(ctx context.Context, c *assert.CollectT, cl client.Client) {
+		gw := getGatewayByLabelSelector(gatewayLabelSelector, ctx, c, cl)
+		controlplanes, err := gateway.ListControlPlanesForGateway(ctx, cl, gw)
+		if err != nil {
+			c.Errorf("failed to list ControlPlanes for Gateway %q: %v", client.ObjectKeyFromObject(gw), err)
+			c.FailNow()
+		}
+		require.Len(c, controlplanes, 1)
```
Comment (pmalek) [inline test/e2e/test_helm_install_upgrade.go]: Given that those assertions are run inside the `require.EventuallyWithT` I believe we want to use `assert` + return instead of `require` so that we don't abort the whole test right away but we continue to try to satisfy the assertion.

File: test/e2e/test_helm_install_upgrade.go [hunk 6]
```diff
@@ -178,6 +179,12 @@ func TestHelmUpgrade(t *testing.T) {
 						gatewayDataPlaneDeploymentIsNotPatched("gw-upgrade-123-current=true")(ctx, c, cl.MgrClient)
 					},
 				},
+				{
+					Name: "Cluster wide resources owned by the ControlPlane get the proper set of labels",
```
Comment (pmalek) [inline test/e2e/test_helm_install_upgrade.go:183]: 👍

File: pkg/utils/kubernetes/owners.go
Comment (pmalek) [inline pkg/utils/kubernetes/owners.go:1]: 👍

File: pkg/utils/kubernetes/owners.go [hunk 7]
```diff
@@ -38,6 +42,46 @@ func SetOwnerForObject(obj, owner client.Object) {
 	}
 }
 
+// managingObjectT is an interface that is used to represent a managing object.
+// managingObjects can be of type Gateway, ControlPlane, or DataPlane.
```
Comment (pmalek) [inline pkg/utils/kubernetes/owners.go]: Nit: While this uses the `interface` keyword this is a [type constraint](https://go.dev/blog/intro-generics#type-sets).

I'd reword this as:

```suggestion
// managingObjectT is type constraint that is used to represent a managing object.
// Currently it can be one of: Gateway, ControlPlane, or DataPlane.
```

**Review by pmalek — COMMENTED** [review]
I'd add a Changelog entry describing this change. WDYT?

**Comment by mlavacca** [comment]
> I'd add a Changelog entry describing this change. WDYT?

I don't see any new commit 🤔 I've created a CHANGELOG entry though

File: pkg/utils/kubernetes/ownerrefs.go
Comment (pmalek) [reply pkg/utils/kubernetes/ownerrefs.go:1]: 👍

**Review by pmalek — APPROVED** [review]
I believe this is good to go. 

Thanks for working on this @mlavacca 🙇
